src/main/java/de/tum/in/test/api/util/ClassMemberAccessor.java

```diff
@@ -0,0 +1,157 @@
+					// searching for private fields to be able to provide error messages to the
+					// users that the field exists, but with the wrong visibility.
+					return getInheritedField(declaringClass, c, fieldName).stream();
+				} else {
+					return Stream.of(c.getField(fieldName));
+				}
+			} catch (NoSuchFieldException nsfe) {
+				return Stream.empty();
+			}
+		}).findFirst().orElseThrow(() -> new NoSuchFieldException(fieldName));
+	}
+
+	/**
+	 * Searches for a field in the target class that might be inherited from the
+	 * declaring class.
+	 *
+	 * @param targetClass    The class in which the field should be accessible.
+	 * @param declaringClass The class from which the field might be inherited.
+	 * @param fieldName      The name of the attribute.
+	 * @return A field that is accessible in the target class.
+	 * @throws NoSuchFieldException Thrown if no field with the given name could be
+	 *                              found in either class.
+	 */
+	private static Optional<Field> getInheritedField(Class<?> targetClass, Class<?> declaringClass, String fieldName)
+			throws NoSuchFieldException {
+		Field field = declaringClass.getDeclaredField(fieldName);
+		if (isInherited(targetClass, declaringClass, field.getModifiers())) {
+			return Optional.of(field);
+		} else {
+			return Optional.empty();
+		}
+	}
+
+	/**
+	 * Searches for all classes and interfaces the given class is inheriting from.
+	 * <p>
+	 * The given class itself is part of the result. Performs a depth-first search
+	 * starting with classes, then interfaces.
+	 *
+	 * @param clazz The class for which the inheritance tree should be traversed.
+	 * @return A stream of the given class and all superclasses and interfaces it
+	 *         inherits from.
+	 */
+	private static Stream<Class<?>> getClassHierarchy(Class<?> clazz) {
+		Stream<Class<?>> directSuperclasses = Stream
+				.concat(Stream.of(clazz.getSuperclass()), Arrays.stream(clazz.getInterfaces()))
+				.filter(Objects::nonNull);
+		return Stream.concat(Stream.of(clazz), directSuperclasses.flatMap(ClassMemberAccessor::getClassHierarchy));
+	}
+
+	/**
+	 * Checks if a field or method member of a class is accessible in a subclass.
+	 *
+	 * @param targetClass     The class in which the class member of
+	 *                        {@code declaredInClass} should be accessible. Assumes
+	 *                        that {@code declaredInClass} is a superclass of
+	 *                        {@code targetClass}.
+	 * @param declaredInClass The class in which the member was declared.
+	 * @param modifier        The modifiers of the member.
+	 * @return True, if the class member of {@code declaredInClass} is accessible in
+	 *         its subclass {@code targetClass}.
+	 */
+	private static boolean isInherited(Class<?> targetClass, Class<?> declaredInClass, int modifier) {
+		return targetClass.equals(declaredInClass) || !Modifier.isPrivate(modifier);
```

Comment:
There could be a missed special case here:

protected grants more visibility than package-private, giving subclasses outside the current package access to the member. With package-private, a subclass residing in another package won't have access. But this method will return true in this case (not private).

Reply:
Good catch. In c4e3cffc05a925e0bb2adb324ba8d11265a92552 this should now be handled correctly.

The two tests `getProtectedMethodAccessInSubpackage` and `getPackagePrivateMethodNoAccessInSubpackage` are new, the remaining ones are unchanged but just got reordered.